src/sentry/models/notificationaction.py

```diff
@@ -142,6 +149,39 @@ class Meta:
         db_table = "sentry_notificationactionproject"
 
 
+class ActionRegistration(metaclass=ABCMeta):
+    def __init__(self, action: NotificationAction):
+        self.action = action
+
+    @abstractmethod
+    def fire(self, data: Any) -> None:
+        """
+        Handles delivering the message via the service from the action and specified data.
+        """
+        pass
+
+    def validate_action(self, data: NotificationActionInputData) -> None:
```

Comment:
will we just check if this function is implemented before using it instead of `NotificationActionSerializer`?

Reply:
It's a bit different, I added a change to the serializer as per my TODO, but the tl;dr is that we're just running these functions as an additional validation step. If these functions raise validation errors, they'll stop the POST/PUT (see new tests)